### PR TITLE
libcephfs: add ceph_p{read,write}v

### PR DESCRIPTION
**Pull request by JevonQ**
Add interfaces in libcephfs to enable users to read/write data via iovec structure.

Signed-off-by: Jevon Qiao <qiaojianfeng@unitedstack.com>

### CODE REVIEW

**Comment by JevonQ** [comment]
It's necessary to have the preadv/pwritev interfaces in libcephfs so that the consumers will be able to read/write data via iovec structure. With them, it will make libcephfs work seamlessly with the consumers having the ability to gather IOs as a iovec.

This commit is the one I cooked up and I'm looking for the feedback. Any comment would be highly appreciated.

Thanks,
Jevon

**Comment by tchaikov** [comment]
@JevonQ i am not the best person to comment on your new libceph APIs. but one thing you might want to add to this pull request is some tests in src/test/libcephfs/test.cc.

**Comment by JevonQ** [comment]
@tchaikov Really thank you for the prompt response. I'm new to ceph and not quite familiar with the process. I'll add some tests in test.cc as you mentioned above, thank you for the advice.

File: src/client/Client.cc [hunk 1]
```diff
@@ -7659,6 +7720,9 @@ int Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf)
   }
 
   put_cap_ref(in, CEPH_CAP_FILE_WR);
+  if (buf == NULL && iov != NULL) {
+    delete[] bparr;
```
Comment (jcsp) [inline src/client/Client.cc]: If you initialize bparr to null where it's declared, you can avoid the need for a conditional here

File: src/client/Client.cc [hunk 2]
```diff
@@ -7508,10 +7559,20 @@ int Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf)
...
+  bufferptr *bparr;
+  if (buf) {
+      bufferptr bp;
+      if (size > 0) bp = buffer::copy(buf, size);
+      bl.push_back( bp );
+  } else if (iov){
+      //iov case 
+      bparr = new bufferptr[iovcnt];
+      for (int i = 0; i < iovcnt; i++) {
+        if (iov[i].iov_len > 0) bparr[i] = buffer::copy((char*)iov[i].iov_base, iov[i].iov_len);
```
Comment (jcsp) [inline src/client/Client.cc]: Avoid single line if statements (only allowed for for very short statements)
(https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Conditionals)

File: src/client/Client.cc [hunk 3]
```diff
@@ -7450,13 +7451,63 @@ int Client::write(int fd, const char *buf, loff_t size, loff_t offset)
...
+    if (fh->flags & O_PATH)
+        return -EBADF;
+#endif
+    loff_t totallen = 0;
+    for (int i = 0; i < iovcnt; i++) {
+        totallen += iov[i].iov_len;
+    }
+    if (write) {
+        int w = _write(fh, offset, totallen, NULL, iov, iovcnt);
+        ldout(cct, 3) << "write(" << fd << ", \"...\", " << totallen << ", " << offset << ") = " << w << dendl;
```
Comment (jcsp) [inline src/client/Client.cc]: Would be nice to make this debug message distinguishable from normal writes by just making it "writev(..." instead (same for the read one)

File: src/client/Client.cc [hunk 4]
```diff
@@ -7450,13 +7451,63 @@ int Client::write(int fd, const char *buf, loff_t size, loff_t offset)
...
+        return w;
+    } else {
+        bufferlist bl;
+        int r = _read(fh, offset, totallen, &bl);
+        ldout(cct, 3) << "read(" << fd << ", " <<  offset << ") = " << r << dendl;
+        int bufoff = 0;
+        uint32_t rlen = 0;
+        for (int j = 0, tmplen = r; j < iovcnt && tmplen > 0; j++) {
```
Comment (jcsp) [inline src/client/Client.cc]: more descriptive name for tmplen might "remaining"

**Comment by jcsp** [comment]
Looks functional to me, but will be more sure when we have some tests :-)

**Comment by gregsfortytwo** [comment]
Thanks for taking this one. :) Some tests have appeared!

**Comment by JevonQ** [comment]
@jcsp Thank you very much for your comments, I'll address them right away. @jcsp @gregsfortytwo @tchaikov I've already added a test for these two new interfaces and just committed the code changes a few minutes ago, please have a look.

Thanks,
Jevon

File: src/client/Client.cc (hunk 1 above)
Comment (JevonQ) [inline src/client/Client.cc]: Make sense, I'll revise the code.

**Comment by yuyuyu101** [comment]
@JevonQ Cool, well done!!

File: src/client/Client.cc (hunk 1 above)
Comment (JevonQ) [inline src/client/Client.cc]: Make sense, just a habit built from writing c code.

File: src/client/Client.cc (hunk 2 above)
Comment (JevonQ) [inline src/client/Client.cc]: Got it. BTW, is there any tool available for me to run the style check against my changes? Is it a must have step before integrating the code into the upstream ceph repo?

File: src/client/Client.cc (hunk 3 above)
Comment (JevonQ) [inline src/client/Client.cc]: Cannot agree more.

**Comment by JevonQ** [comment]
@yuyuyu101 Made the first step.

**Comment by JevonQ** [comment]
@jcsp Your comments listed above have been addressed, please have a check.

**Comment by jcsp** [comment]
Merged, thanks!

**Comment by JevonQ** [comment]
@jcsp Thank you for the effort.

**Pull request merged**